### ISSUE TITLE
Implementation of Bayesian Personalized Ranking Loss

### ISSUE THREAD

**Issue by LukasHaas** [pytorch/pytorch#69858]
### 🚀 The feature, motivation and pitch

I am currently working on graph neural network applications to recommender systems. A common loss function that keeps coming is the Bayesian Personalized Ranking loss as described in the 2012 paper by Rendle et al. titled [BPR: Bayesian Personalized Ranking from Implicit Feedback](https://arxiv.org/abs/1205.2618). 

Given that the BPR loss function is a useful contrastive loss to train recommender systems, I implemented it in PyTorch and would now love to submit it for review.

### Alternatives

While there are other ranking loss functions, the BPR loss would be a valuable addition to the PyTorch library.

### Additional context

One example of a domain and model for which the BPR loss would be helpful is music playlist continuation. I recently implemented the LightGCN graph neural network from the paper [LightGCN: Simplifying and Powering Graph Convolution Network for Recommendation](https://arxiv.org/abs/2002.02126) to recommend new songs for playlists which uses the BPR loss function to learn embeddings for nodes in a graph.

I have working code for the BPR loss ready to be submitted.

**Comment by mruberry** [comment]
Thank you for the request, @LukasHaas.

Google Scholar indicates this paper has 3916 citations and arxiv says there are 20 public implementations of it, including at least one PyTorch implementation (https://github.com/sh0416/bpr). Seems popular enough to look at incorporating into PyTorch core.

I think we would accept a PR adding this

**Comment by LukasHaas** [comment]
This is great to hear, will get working on it.

**Comment by daviddavo** [comment]
Doesn't BPR Loss need the positive and negative samples too? The forward method would have quite a different interface to the other methods in the loss package.

**Comment by AvishekMondalQC** [comment]
I think the main challenge here would be vectorizing the selection of the negative examples. 

Specifically, I'm referring to this snippet of code in the PyTorch implementation referenced in https://github.com/pytorch/pytorch/issues/69858#issuecomment-996202985: 

```python
    def _example(self, idx):
        u = self.pair[idx][0]
        i = self.pair[idx][1]
        j = np.random.randint(self.num_item)
        while j in self.user_list[u]:
            j = np.random.randint(self.num_item)
        return u, i, j
```

I'm not sure how you'd have an interface that supports something like this?

**Comment by daviddavo** [comment]
You could generate a list of N random numbers where N is the number of samples, then you get the indices where "false negatives are", and generate another set random numbers, and combine the two arrays with the indices.

Pseudocode:
```
N = 32 # Number of items

# assert that there is at least a possible
# negative to avoid infinite loop

neg = np.random.randint(N)
pos_idx = get_pos_idx(...)
while len(pos_idx)>0:
  neg[pos_idx] = np.random.randint(len(pos_idx))
  pos_idx = get_pos_idx(...)
```
